lib/virt_autotest/transactional_utils.pm

```diff
@@ -0,0 +1,95 @@
+# SUSE's openQA tests
+#
+# Copyright 2022 SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+# Summary: Virtualization test utilities for transactional server
+# Maintainer: Wayne Chen <wchen@suse.com>
+
+package virt_autotest::transactional_utils;
+
+use base Exporter;
+use Exporter;
+
+use strict;
+use warnings;
+use utils;
+use version_utils;
+use testapi;
+use Utils::Architectures;
+use Carp;
+use transactional;
+
+our @EXPORT = qw(enter_trup_shell quit_trup_shell reboot_into_new_snapshot);
+
+=head2 enter_trup_shell
+
+  enter_trup_shell(global_options => $global_options, shell_options => $shell_options)
+
+Enter into transactional update shell by entering command on transactional server:
+transactional-update $global_options shell $shell_options. The two arguments for
+this subroutine, global_options and shell_options, are all text strings that are
+composed of space separated options for transactional-update and shell respectively.
+
+=cut
+
+sub enter_trup_shell {
+    my (%args) = @_;
+
+    $args{global_options} //= '';
+    $args{shell_options} //= '';
+    enter_cmd("transactional-update $args{global_options} shell $args{shell_options}; echo trup_shell-status-\$? > /dev/$serialdev");
+    save_screenshot;
+    wait_still_screen;
+    save_screenshot;
+}
+
+=head2 quit_trup_shell
+
+  quit_trup_shell(change_expected => $change_expected)
+
+Quit transactional update shell by entering exit. Check if any changes that request
+reboot to take effect. The only argument change_expected is passed to and going to
+be used by subroutine check_reboot_changes. Please refer to the lib/transactional.pm
+for more information.
+
+=cut
+
+sub quit_trup_shell {
+    my (%args) = @_;
+
+    $args{change_expected} //= 1;
+    enter_cmd("exit");
+    save_screenshot;
+    wait_serial('trup_shell-status-0') || die "'transactional-update shell' didn't finish";
+    save_screenshot;
+    assert_script_run("transactional-update status");
+    save_screenshot;
+    reboot_into_new_snapshot();
+}
+
+=head2 reboot_into_new_snapshot
+
+  reboot_into_new_snapshot
+
+Check whether new snapshot is generated and reboot into this new snapshot if there
+are changes happened.
+
+=cut
+
+sub reboot_into_new_snapshot {
```

Comment:
check_reboot_changes will terminate test run if there is no change(snapshot) happened

Reply:
I read the code again, I am afraid it can be controlled by $change_expected when calling the function? Besides, if you want to skip this check which can result in die sometimes, you can also introduce a new parameter to the function, eg  skip_expectation_check,   which by default is the old behavior(skip_expectation_check=off), but in your calling, call it with value `on`?

Reply:
I will try to reuse check_reboot_changes later and switch to use it if OK.

Reply:
Okay.